spotguide/spotguide.go

```diff
@@ -0,0 +1,323 @@
+	return "spotguide_repos"
+}
+
+func (s *Repo) AfterFind() error {
+	return yaml.Unmarshal(s.SpotguideRaw, &s.Spotguide)
+}
+
+type LaunchRequest struct {
+	SpotguideName    string   `json:"spotguideName"`
+	RepoOrganization string   `json:"repoOrganization"`
+	RepoName         string   `json:"repoName"`
+	Secrets          []Secret `json:"secrets"`
+}
+
+type Secret struct {
+	Name string `json:"name"`
+}
+
+func (r LaunchRequest) RepoFullname() string {
+	return r.RepoOrganization + "/" + r.RepoName
+}
+
+func getUserGithubToken(userID uint) (string, error) {
+	token, err := auth.TokenStore.Lookup(fmt.Sprint(userID), auth.GithubTokenID)
+	if err != nil {
+		return "", err
+	}
+	return token.Value, nil
+}
+
+func newGithubClientForUser(userID uint) (*github.Client, error) {
+	accessToken, err := getUserGithubToken(userID)
+	if err != nil {
+		return nil, err
+	}
+
+	return newGithubClient(accessToken), nil
+}
+
+func newGithubClient(accessToken string) *github.Client {
+	httpClient := oauth2.NewClient(
+		ctx,
+		oauth2.StaticTokenSource(&oauth2.Token{AccessToken: accessToken}),
+	)
+
+	return github.NewClient(httpClient)
+}
+
+func ScrapeSpotguides() error {
+
+	db := config.DB()
+
+	githubClient := newGithubClient(viper.GetString("github.token"))
+
+	repositories, _, err := githubClient.Repositories.ListByOrg(ctx, SpotguideGithubOrganization, &github.RepositoryListByOrgOptions{
+		ListOptions: github.ListOptions{PerPage: 200},
+	})
+
+	if err != nil {
+		return err
+	}
+
+	for _, repository := range repositories {
+		for _, topic := range repository.Topics {
+			if topic == SpotguideGithubTopic {
+				owner := repository.GetOwner().GetLogin()
+				name := repository.GetName()
+
+				reader, err := githubClient.Repositories.DownloadContents(ctx, owner, name, SpotguidePath, nil)
+				if err != nil {
+					return err
+				}
+
+				spotguideRaw, err := ioutil.ReadAll(reader)
+				if err != nil {
+					return err
+				}
+
+				model := Repo{
+					Name:         repository.GetFullName(),
+					SpotguideRaw: spotguideRaw,
+				}
+
+				err = db.Where(&model).Assign(&model).FirstOrCreate(&Repo{}).Error
+
+				if err != nil {
+					return err
+				}
+
+				break
+			}
+		}
+	}
+
+	return nil
+}
+
+func GetSpotguides() ([]*Repo, error) {
+	db := config.DB()
+	spotguides := []*Repo{}
+	err := db.Find(&spotguides).Error
+	return spotguides, err
+}
+
+func GetSpotguide(name string) (*Repo, error) {
+	db := config.DB()
+	spotguide := Repo{}
+	err := db.Where("name = ?", name).Find(&spotguide).Error
+	return &spotguide, err
+}
+
+// curl -X POST -H "Authorization: Bearer $TOKEN" -H "Content-Type: application/json" -v http://localhost:9090/api/v1/orgs/1/spotguides -d '{"repoName":"spotguide-test", "repoOrganization":"banzaicloud"}'
+func LaunchSpotguide(request *LaunchRequest, orgID, userID uint) error {
+
+	err := createGithubRepo(request, userID)
+	if err != nil {
+		log.Errorln("Failed to create GitHub repository", err.Error())
```

Comment:
consider using the pkg/ errors package here. we can call errors.cause before sending back bad requests to the customer.